src/Microsoft.AspNet.OData.Shared/Builder/EdmModelHelperMethods.cs

```diff
@@ -638,13 +652,64 @@ private static void AddCoreVocabularyAnnotations(this EdmModel model, IEnumerabl
             if (edmProperties.Any())
             {
                 IEdmCollectionExpression collectionExpression = new EdmCollectionExpression(edmProperties.Select(p => new EdmPropertyPathExpression(p.Name)).ToArray());
-                IEdmTerm term = Microsoft.OData.Edm.Vocabularies.V1.CoreVocabularyModel.ConcurrencyTerm;
+                IEdmTerm term = CoreVocabularyModel.ConcurrencyTerm;
                 EdmVocabularyAnnotation annotation = new EdmVocabularyAnnotation(target, term, collectionExpression);
                 annotation.SetSerializationLocation(model, EdmVocabularyAnnotationSerializationLocation.Inline);
                 model.SetVocabularyAnnotation(annotation);
             }
         }
 
+        private static void AddDerivedTypeConstraintAnnotationForProperties(this EdmModel model, EdmTypeMap edmTypeMap)
+        {
+            Dictionary<IEdmProperty, PropertyConfiguration>
+                propertyConfiguration = edmTypeMap.EdmPropertyConfigurations;
+
+            foreach (var pair in propertyConfiguration)
+            {
+                if (pair.Value.DerivedTypeConstraints.Count > 0 )
+                {
+                    model.AddDerivedTypeConstraintAnnotation(pair.Key, edmTypeMap.EdmTypes, pair.Value.DerivedTypeConstraints.ToList());
+                }
+            }         
+        }
+
+        private static void AddDerivedTypeConstraintAnnotationForNavigationSource(this EdmModel model, IEdmVocabularyAnnotatable target,
+            NavigationSourceConfiguration navigationSourceConfiguration, EdmTypeMap edmTypeMap)
+        {
+            if (navigationSourceConfiguration != null)
+            {
+                model.AddDerivedTypeConstraintAnnotation(target,
+                    edmTypeMap.EdmTypes, navigationSourceConfiguration.DerivedTypeConstraints.ToList());
+            }
+        }
+
+        internal static void AddDerivedTypeConstraintAnnotation(this EdmModel model, IEdmVocabularyAnnotatable target, 
+            Dictionary<Type, IEdmType> edmTypeMap, List<Type> types)
+        {
+            if (types.Count == 0)
+            {
+                return;
+            }
+
+            IEdmTerm term = ValidationVocabularyModel.DerivedTypeConstraintTerm;
+            List<EdmStringConstant> collectionConstants = new List<EdmStringConstant>();
+            foreach (var type in types)
+            {
+                if (edmTypeMap.ContainsKey(type))
+                {
+                    collectionConstants.Add(new EdmStringConstant(edmTypeMap[type].FullTypeName()));
+                }
+                else
+                {
+                    throw Error.InvalidOperation(SRResources.CantFindEdmType, type);
+                }
+            }
+
+            var collectionExpression = new EdmCollectionExpression(collectionConstants);
+            EdmVocabularyAnnotation annotation = new EdmVocabularyAnnotation(target, term, collectionExpression);
+            model.SetVocabularyAnnotation(annotation);
```

Comment:
>SetVocabularyAnnotation [](start = 18, length = 23)

I would generally think (at least by default) we should add the derived type constraint to the model element itself, not the model.

Reply:
The model element is specified by the target in the annotation. As per my discussion with Sam, it seemed like the right thing to do. We can sync up if I misunderstood something.

Reply:
That's fine; I just want to make sure that this constraint, in particular, is written inline by default.

---
In reply to: [383581324](https://github.com/OData/WebApi/pull/1977#discussion_r383581324) [](ancestors = 383581324)